pkg/pillar/types/global.go

```diff
@@ -103,312 +105,408 @@ func (gs *GlobalStatus) UpdateItemValuesFromGlobalConfig(gc GlobalConfig) {
+}
+
+func (specMap *ConfigItemSpecMap) AddAgentSettingStringItem(key AgentSettingKey, defaultString string) {
+	configItem := ConfigItemSpec{
+		ItemType:      ConfigItemTypeString,
+		Key:           string(key),
+		StringDefault: defaultString,
 	}
-	if newgc.MintimeUpdateSuccess < GlobalConfigMinimums.MintimeUpdateSuccess {
-		log.Warnf("Enforce minimum MintimeUpdateSuccess received %d; using %d",
-			newgc.MintimeUpdateSuccess, GlobalConfigMinimums.MintimeUpdateSuccess)
-		newgc.MintimeUpdateSuccess = GlobalConfigMinimums.MintimeUpdateSuccess
+	specMap.AgentSettings[key] = configItem
+	log.Debugf("Added string item %s", key)
+}
+
+func (specMap *ConfigItemSpecMap) ParseItem(configMap *ConfigItemValueMap, value string, key string) error {
+	// legacy per-agent setting key debug.<agentname>.xxx
+	// new per-agent setting key agent.<agentname>.debug.xxx
+	var configSpec ConfigItemSpec
+	agentSetting := false
+	globalSetting := false
+	components := strings.Split(key, ".")
+	if strings.HasPrefix(key, "agent") || strings.HasPrefix(key, "debug") {
+		if strings.HasPrefix(key, "agent") && len(components) > 2 {
+			components = components[2:]
+		} else if strings.HasPrefix(key, "debug") && len(components) > 2 {
+			components[1] = ""
+		} else {
+			return fmt.Errorf("Unable to find agent name for per-agent setting. Key: %s", key)
+		}
+		key = strings.Join(components, ".")
+		agentSetting = true
+	} else if _, ok := specMap.GlobalSettings[GlobalSettingKey(key)]; ok {
+		globalSetting = true
+	} else {
+		return fmt.Errorf("Item is neither a global nor a per-agent setting. Key: %s", key)
+	}
+	if agentSetting {
+		configSpec = specMap.AgentSettings[AgentSettingKey(key)]
+		agentName := components[1]
+		val, err := parseConfigItemToValue(configSpec, value)
+		if err == nil {
+			configMap.AgentSettings[agentName][AgentSettingKey(key)] = val
+		} else {
+			return err
+		}
+	} else if globalSetting {
+		configSpec = specMap.GlobalSettings[GlobalSettingKey(key)]
+		val, err := parseConfigItemToValue(configSpec, value)
+		if err == nil {
+			configMap.GlobalSettings[GlobalSettingKey(key)] = val
+		} else {
+			return err
+		}
+	}
+	return nil
+}
+
+type ConfigItemValue struct {
+	Key      string
+	ItemType ConfigItemType
+
+	intValue      uint32
+	stringValue   string
+	boolValue     bool
+	triStateValue TriState
+}
+
+type ConfigItemValueMap struct {
+	// GlobalSettings - Map Key: GlobalSettingKey, ConfigItemValue.Key: GlobalSettingKey
+	GlobalSettings map[GlobalSettingKey]ConfigItemValue
+	// AgentSettings - Map Outer Key: agentName, Map Inner Key: AgentSettingKey ConfigItemValue.Key: AgentSettingKey
+	AgentSettings map[string]map[AgentSettingKey]ConfigItemValue
+}
+
+func (configPtr *ConfigItemValueMap) globalConfigItemValue(key GlobalSettingKey) ConfigItemValue {
+	specMapPtr := NewConfigItemSpecMap()
+	val, okVal := configPtr.GlobalSettings[key]
+	spec, _ := specMapPtr.GlobalSettings[key]
+	if okVal {
+		return val
+	} else {
```

Comment:
is it better to return an error here too?? 
   func (configPtr *ConfigItemValueMap) globalConfigItemValue(key GlobalSettingKey) (ConfigItemValue, error) ??